presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -49,34 +56,172 @@ public JoinStatsRule(FilterStatsCalculator filterStatsCalculator)
+            else {
+                // either leftNDV or rightNDV is NaN
+                return UNKNOWN_STATS;
+            }
+        }
+
+        return outputStats;
     }
 
-    private PlanNodeStatsEstimate crossJoinStats(JoinNode joinNode, Lookup lookup, Session session, Map<Symbol, Type> types)
+    @VisibleForTesting
+    PlanNodeStatsEstimate addAntiJoinStats(PlanNodeStatsEstimate joinStats, PlanNodeStatsEstimate antiJoinStats, Set<Symbol> joinSymbols)
     {
-        PlanNodeStatsEstimate leftStats = lookup.getStats(joinNode.getLeft(), session, types);
-        PlanNodeStatsEstimate rightStats = lookup.getStats(joinNode.getRight(), session, types);
+        checkState(joinStats.getSymbolsWithKnownStatistics().containsAll(antiJoinStats.getSymbolsWithKnownStatistics()));
+
+        double joinOutputRowCount = joinStats.getOutputRowCount();
+        double antiJoinOutputRowCount = antiJoinStats.getOutputRowCount();
+        double totalRowCount = joinOutputRowCount + antiJoinOutputRowCount;
+        PlanNodeStatsEstimate outputStats = joinStats.mapOutputRowCount(rowCount -> rowCount + antiJoinOutputRowCount);
+
+        for (Symbol symbol : antiJoinStats.getSymbolsWithKnownStatistics()) {
+            outputStats = outputStats.mapSymbolColumnStatistics(symbol, joinColumnStats -> {
+                SymbolStatsEstimate antiJoinColumnStats = antiJoinStats.getSymbolStatistics(symbol);
+                // weighted average
+                double newNullsFraction = (joinColumnStats.getNullsFraction() * joinOutputRowCount + antiJoinColumnStats.getNullsFraction() * antiJoinOutputRowCount) / totalRowCount;
+                double distinctValues;
+                if (joinSymbols.contains(symbol)) {
+                    distinctValues = joinColumnStats.getDistinctValuesCount() + antiJoinColumnStats.getDistinctValuesCount();
+                }
+                else {
+                    distinctValues = joinColumnStats.getDistinctValuesCount();
+                }
+                return SymbolStatsEstimate.buildFrom(joinColumnStats)
+                        .setLowValue(Math.min(joinColumnStats.getLowValue(), antiJoinColumnStats.getLowValue()))
+                        .setHighValue(Math.max(joinColumnStats.getHighValue(), antiJoinColumnStats.getHighValue()))
+                        .setDistinctValuesCount(distinctValues)
+                        .setNullsFraction(newNullsFraction)
+                        .build();
+            });
+        }
 
+        // add nulls to columns that don't exist in right stats
+        for (Symbol symbol : difference(joinStats.getSymbolsWithKnownStatistics(), antiJoinStats.getSymbolsWithKnownStatistics())) {
+            outputStats = outputStats.mapSymbolColumnStatistics(symbol, joinColumnStats ->
+                    SymbolStatsEstimate.buildFrom(joinColumnStats)
+                            .setNullsFraction((joinColumnStats.getNullsFraction() * joinOutputRowCount + antiJoinOutputRowCount) / totalRowCount)
```

Comment:
there should be `mapNullFraction` method in `SymbolStatsEstimate` which you could use here.